Implemented notification system for investors subscribing to startups

DimaMakovskyi opened the pull request:
Implemented notification system for investors subscribing to startups

mehalyna commented:
The failing job is caused by issues in notification creation for follows in your tests. Specifically:

- test_follow_creates_notification_for_startup_owner is failing because a Notification is not being created when an investor follows a startup.
- test_duplicate_follow_does_not_create_duplicate_notification is failing because no Notification is created at all (expected: 1, found: 0).

Both failures indicate that the Notification creation logic is either not triggered or broken when a SavedStartup is created.

Solution:
1. Check the code responsible for creating the Notification when a SavedStartup is created (likely in a signal handler).
2. Ensure the signal is properly connected and that the Notification is only created if one does not already exist for this investor/startup pair.

Example fix for your signal (pseudo-code):

```python
from django.db.models.signals import post_save
from django.dispatch import receiver
from .models import SavedStartup, Notification

@receiver(post_save, sender=SavedStartup)
def create_follow_notification(sender, instance, created, **kwargs):
    if created:
        # Ensure not duplicate
        if not Notification.objects.filter(
            user=instance.startup.user,
            notification_type__code="startup_followed",
            related_startup_id=instance.startup.id
        ).exists():
            Notification.objects.create(
                user=instance.startup.user,
                notification_type=...,
                related_startup_id=instance.startup.id,
                message=f"{instance.investor.user.first_name} followed your startup"
            )
```

- Confirm that your signal is imported and registered in AppConfig.
- Run your tests after this change.

See the test file for context: tests/communications/test_notification_triggers.py (ref: c25f73f944161ef8261d88f62067a244b9ee7638).

DimaMakovskyi commented:
feat: add API endpoints for managing saved startups

- Added endpoint to list all startups an investor has saved (GET /api/investor/saved-startups).
- Implemented filtering options to sort or search saved startups.
- Added endpoint to unfollow a startup (DELETE /api/startups/{startup_id}/unsave).
- Ensured authentication and authorization checks are in place.

DimaMakovskyi commented on communications/signals.py:
```diff
@@ -33,37 +45,124 @@ def create_initial_notification_types(sender, **kwargs):
+                triggered_by_type=NotificationTrigger.INVESTOR,
+                priority=NotificationPriority.LOW,
+                related_startup_id=getattr(startup, "id", None),
+                action_link=f"/startups/{getattr(startup, 'id', '')}/followers",
```
fix
